### PR TITLE
1949 Add unit tests to SharedKeychainSessionStorageFactory & User

### PR DESCRIPTION
**Pull request by zamorastp**

### CODE REVIEW

File: Projects/AccountSDKIOSWeb/Tests/AccountSDKIOSWebTests/SimplifiedLogin/SharedKeychainSessionStorageFactoryTests.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        
+        let keychain = SharedKeychainSessionStorageFactory(keychain: keychainSessionStorageMock, sharedKeychain: sharedKeychainSessionStorageMock).makeKeychain(clientId: "client_id", service: "service_name", accessGroup: accessGroup, appIdentifierPrefix: "AZWSDFGHIJ")
+        
+        XCTAssertIdentical(keychain, sharedKeychainSessionStorageMock)
+        verify(sharedKeychainSessionStorageMock).checkEntitlements()
+        verify(sharedKeychainSessionStorageMock).getAll()
+        verify(sharedKeychainSessionStorageMock, never()).store(any(), accessGroup: sharedAccessGroup, completion: anyClosure())
+    }
+    
+    func testReturnsSharedKeychainWithUpdatedAccessGroupForItem() {
```
Comment (Daniel-Echegaray) [inline Projects/AccountSDKIOSWeb/Tests/AccountSDKIOSWebTests/SimplifiedLogin/SharedKeychainSessionStorageFactoryTests.swift:66]: The name of this function makes me think that this unit test is for assuring that the SharedKeychain gets its accessGroup updated.

But on diving in to the code a bit I realise the test is to assure that the sharedKeychain overwrite pre-existing userSession

**Review by Daniel-Echegaray — APPROVED** [review]
I had to re-remember how the `SharedKeychainSessionStorageFactoryTests`.
Perhaps the unit test names could be made a bit clearer. 
But all in all looks good.
I added some comments

File: Projects/AccountSDKIOSWeb/Tests/AccountSDKIOSWebTests/SimplifiedLogin/SharedKeychainSessionStorageFactoryTests.swift [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import XCTest
+import Cuckoo
+@testable import AccountSDKIOSWeb
+
+final class SharedKeychainSessionStorageFactoryTests: XCTestCase {
+    
+    let accessGroup = "com.foo.bar"
+    let sharedAccessGroup = "AZWSDFGHIJ.com.schibsted.simplifiedLogin"
+    
+    func testReturnsAppKeychainWhenNoAppIdentifierIsProvided() {
```
[the remark this answers is not in the document]
Comment (zamorastp) [reply Projects/AccountSDKIOSWeb/Tests/AccountSDKIOSWebTests/SimplifiedLogin/SharedKeychainSessionStorageFactoryTests.swift:10]: No, it's good. When we use a standard keychain (not shared) and any custom access group was not provided then this property is nil

File: Sources/AccountSDKIOSWeb/Lib/Storage/Keychain/SharedKeychainSessionStorageFactory.swift [hunk 3]
```diff
@@ -3,10 +3,17 @@ import Foundation
 struct SharedKeychainSessionStorageFactory {
     
     static let sharedKeychainGroup = "com.schibsted.simplifiedLogin"
+    private var keychain: KeychainSessionStorage?
+    private var sharedKeychain: KeychainSessionStorage?
     
-    static func makeKeychain(clientId: String, service: String, accessGroup: String? = nil, appIdentifierPrefix: String? = nil) -> KeychainSessionStorage {
+    init(keychain: KeychainSessionStorage? = nil, sharedKeychain: KeychainSessionStorage? = nil) {
```
[the remark this answers is not in the document]
Comment (zamorastp) [reply Sources/AccountSDKIOSWeb/Lib/Storage/Keychain/SharedKeychainSessionStorageFactory.swift:9]: Yes, it's used only in tests. A modern approach to unit tests is to use dependency injection. You can then simply inject your own mock (which is easy to write and easy to inject). And then no Cuckoo like library (which limits implementation) is needed

[the remark this answers is not in the document]
Comment (zamorastp) [reply Sources/AccountSDKIOSWeb/Lib/Storage/Keychain/SharedKeychainSessionStorageFactory.swift:9]: I have no idea how to test SImplifiedLoginManager now. I cannot mock Client with cuckoo, I cannot write my own client mock (due the fact that i cannot inject dependencies in Client). Like Siimplified Login feature is almost not testable in SDK :(

File: Projects/AccountSDKIOSWeb/Tests/Util/UIWindow+VisibleVCTests.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        navigationController.pushViewController(vc, animated: false)
+        let window = UIWindow(frame: UIScreen.main.bounds)
+        window.rootViewController = navigationController
+        window.makeKeyAndVisible()
+        
+        XCTAssertEqual(window.visibleViewController, vc)
+        
+    }
+    
+    func testReturnsVisibleViewControllerFromTabBar() {
```
[the remark this answers is not in the document]
Comment (zamorastp) [reply Projects/AccountSDKIOSWeb/Tests/Util/UIWindow+VisibleVCTests.swift:29]: Yeah, I can write a test with VC embedded in navigation embedded in tab bar

File: Projects/AccountSDKIOSWeb/Tests/AccountSDKIOSWebTests/SimplifiedLogin/SharedKeychainSessionStorageFactoryTests.swift (hunk 1 above)
Comment (zamorastp) [reply Projects/AccountSDKIOSWeb/Tests/AccountSDKIOSWebTests/SimplifiedLogin/SharedKeychainSessionStorageFactoryTests.swift:66]: No, its not overwriting. This tests sceanrio:
- application has user session in standard keychain
- then app identifier is provided and entitlement is added to the app 
- so on next start i want to use shared keychain in my app but i also want to have my user session - so i migrate it 

this is why i have two mocks of keychain here

File: Sources/AccountSDKIOSWeb/Lib/Storage/Keychain/SharedKeychainSessionStorageFactory.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (Daniel-Echegaray) [reply Sources/AccountSDKIOSWeb/Lib/Storage/Keychain/SharedKeychainSessionStorageFactory.swift:9]: In regards to this specific case of SharedKeychainSessionStorageFactory. Maybe you could have this init as a fileprivate extension in the test target?

About the Client. Isnt it possible to generate a mock for it in the generate_mocks file?

File: Projects/AccountSDKIOSWeb/Tests/AccountSDKIOSWebTests/SimplifiedLogin/SharedKeychainSessionStorageFactoryTests.swift (hunk 1 above)
Comment (Daniel-Echegaray) [reply Projects/AccountSDKIOSWeb/Tests/AccountSDKIOSWebTests/SimplifiedLogin/SharedKeychainSessionStorageFactoryTests.swift:66]: Aha I see 👌 

So its testing migration to SharedKeychain

File: Sources/AccountSDKIOSWeb/Lib/Storage/Keychain/SharedKeychainSessionStorageFactory.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (zamorastp) [reply Sources/AccountSDKIOSWeb/Lib/Storage/Keychain/SharedKeychainSessionStorageFactory.swift:9]: As we talked lately I encountered a lot of problems trying to unit test SimplifiedLoginManager.  Unfortunately, the Client has extensions that make it impossible to properly mock. 
What was your idea to test this class when you write it? I mean unit tests - when you test a specific class with all dependency mocked?

[the remark this answers is not in the document]
Comment (zamorastp) [reply Sources/AccountSDKIOSWeb/Lib/Storage/Keychain/SharedKeychainSessionStorageFactory.swift:9]: Regarding SharedKeychainSessionStorageFactory - this is an internal SDK struct.  Class usage didn't change, it's just extended. You can provide dependencies, or they will be created for you.

**Pull request merged**